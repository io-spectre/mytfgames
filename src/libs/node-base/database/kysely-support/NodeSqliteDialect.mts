/*
The MIT License (MIT)

Copyright (c) 2022 Sami KoskimÃ¤ki

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import {
  DatabaseIntrospector,
  Dialect,
  DialectAdapter,
  Driver,
  Kysely,
  QueryCompiler,
  SqliteAdapter,
  SqliteIntrospector,
  SqliteQueryCompiler,
} from "kysely";

import { NodeSqliteDialectConfig } from "./NodeSqliteConfig.mjs";
import { NodeSqliteDriver } from "./NodeSqliteDriver.mjs";

/**
 * SQLite dialect that uses the [better-sqlite3](https://github.com/JoshuaWise/better-sqlite3) library.
 *
 * The constructor takes an instance of {@link NodeSqliteDialectConfig}.
 *
 * ```ts
 * import Database from 'better-sqlite3'
 *
 * new SqliteDialect({
 *   database: new Database('db.sqlite')
 * })
 * ```
 *
 * If you want the pool to only be created once it's first used, `database`
 * can be a function:
 *
 * ```ts
 * import Database from 'better-sqlite3'
 *
 * new SqliteDialect({
 *   database: async () => new Database('db.sqlite')
 * })
 * ```
 */
export class NodeSqliteDialect implements Dialect {
  readonly #config: NodeSqliteDialectConfig;

  constructor(config: NodeSqliteDialectConfig) {
    this.#config = Object.freeze({ ...config });
  }

  createDriver(): Driver {
    return new NodeSqliteDriver(this.#config);
  }

  createQueryCompiler(): QueryCompiler {
    return new SqliteQueryCompiler();
  }

  createAdapter(): DialectAdapter {
    return new SqliteAdapter();
  }

  createIntrospector(db: Kysely<unknown>): DatabaseIntrospector {
    return new SqliteIntrospector(db);
  }
}
