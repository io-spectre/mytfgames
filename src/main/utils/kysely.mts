import { Kysely, sql, Transaction } from "kysely";

/**
 * Run migration logic in a transaction on a sqlite database.
 *
 * This function will disable foreign key checks, run the migration, and then
 * re-enable foreign key checks. This is necessary because sqlite does not offer
 * adequate ALTER TABLE support and does not support disabling foreign key
 * checks in a transaction.
 *
 * The any types reflect that the database is not in sync with the types of the
 * ORM as per the Kysely documentation.
 *
 * @param db - The database to run the migration on.
 * @param migration - The migration to run.
 */
export async function sqliteSafeMigration(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  db: Kysely<any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  migration: (trx: Transaction<any>) => Promise<void>,
): Promise<void> {
  await sql`PRAGMA foreign_keys=0`.execute(db);
  try {
    await db.transaction().execute(migration);
  } finally {
    await sql`PRAGMA foreign_keys=1`.execute(db);
  }
}
