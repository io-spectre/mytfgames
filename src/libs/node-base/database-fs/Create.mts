import { posix } from "node:path/posix";

import { sql, Transaction } from "kysely";

import {
  AppDatabase,
  AppQueryCreator,
  DatabaseProvider,
  NodeType,
  WellKnownDirectory,
} from "$node-base/database";
import { DbfsErrorCode, makeDbfsError, makeLogicError } from "$pure-base/utils";

interface DbfsMakeDirectoryNodeOptions {
  recursive?: boolean;
}
export async function dbfsMakeDirectoryNode(
  qc: DatabaseProvider,
  path: string,
  options?: DbfsMakeDirectoryNodeOptions,
): Promise<bigint>;
export async function dbfsMakeDirectoryNode(
  qc: DatabaseProvider,
  anchor: bigint,
  path: string,
  options?: DbfsMakeDirectoryNodeOptions,
): Promise<bigint>;
export async function dbfsMakeDirectoryNode(
  qc: DatabaseProvider,
  anchorOrPath: bigint | string,
  pathOrOptions?: string | DbfsMakeDirectoryNodeOptions,
  options?: DbfsMakeDirectoryNodeOptions,
): Promise<bigint> {
  let anchor: bigint;
  let path: string;
  if (typeof anchorOrPath === "string") {
    anchor = WellKnownDirectory.ROOT;
    path = anchorOrPath;
    if (options != null) {
      throw makeLogicError("Unexpected extra options argument");
    }
    if (pathOrOptions == null) {
      options = {};
    } else if (typeof pathOrOptions === "object") {
      options = pathOrOptions;
    } else {
      throw makeLogicError("Unexpected option argument type");
    }
  } else {
    anchor = anchorOrPath;
    if (pathOrOptions == null) {
      throw makeLogicError("Missing path argument");
    } else if (typeof pathOrOptions === "string") {
      path = pathOrOptions;
    } else {
      throw makeLogicError("Unexpected path argument type");
    }
    options ??= {};
  }
  if (anchor !== WellKnownDirectory.ROOT && path.startsWith("/")) {
    throw makeLogicError(
      "Tried to resolve an absolute path with a parent node",
    );
  }

  const fullOptions = {
    recursive: options.recursive ?? false,
  };

  if (qc instanceof Transaction) {
    return dbfsMakeDirectoryNodeImpl(
      qc as Transaction<AppDatabase>,
      anchor,
      path,
      fullOptions,
    );
  }
  return qc.transaction().execute(async (trx) => {
    return dbfsMakeDirectoryNodeImpl(trx, anchor, path, fullOptions);
  });
}

export async function dbfsMakeDirectoryNodeImpl(
  qc: AppQueryCreator,
  parentNo: bigint,
  path: string,
  { recursive }: Required<DbfsMakeDirectoryNodeOptions>,
): Promise<bigint> {
  const normalizedPath = posix.normalize(path);
  const pathSegments = normalizedPath
    .split("/")
    .filter((segment) => segment.length > 0);
  if (pathSegments.length === 0 || normalizedPath === ".") {
    throw makeLogicError("Cannot create a directory without a name");
  } else if (pathSegments[0] === "..") {
    throw makeLogicError("Cannot create a directory above the parent");
  }

  if (pathSegments.length === 1) {
    // fast path for a single segment
    const exists = await qc
      .selectFrom("node_member")
      .innerJoin("node", "node.node_no", "node_member.node_no")
      .select([
        "node_member.node_no",
        "node_member.node_no_parent",
        "node.node_type",
      ])
      .where((eb) =>
        eb.or([
          eb("node_member.node_no", "=", parentNo),
          eb.and([
            eb("node_member.node_no_parent", "=", parentNo),
            eb("node_member.name", "=", pathSegments[0]!),
          ]),
        ]),
      )
      .execute();
    if (exists.length === 0) {
      if (parentNo !== WellKnownDirectory.ROOT) {
        throw makeDbfsError(DbfsErrorCode.ENOENT, "Parent node does not exist");
      }
      // else fall through to create the directory
    } else if (exists.length === 1) {
      if (parentNo === WellKnownDirectory.ROOT) {
        if (exists[0]!.node_type !== NodeType.DIRECTORY) {
          throw makeDbfsError(DbfsErrorCode.EEXIST, "Node is not a directory");
        }
        return BigInt(exists[0]!.node_no);
      }
      if (exists[0]!.node_type !== NodeType.DIRECTORY) {
        throw makeDbfsError(
          DbfsErrorCode.EEXIST,
          "Parent node is not a directory",
        );
      }
      // else fall through to create the directory
    } else if (exists.length === 2) {
      const entry =
        exists[0]!.node_no_parent === parentNo ? exists[0]! : exists[1]!;
      if (entry.node_type !== NodeType.DIRECTORY) {
        throw makeDbfsError(DbfsErrorCode.EEXIST, "Node is not a directory");
      }
      return BigInt(entry.node_no);
    } else {
      throw makeLogicError("Unexpected result from node_member query");
    }
    return dbfsMakeSimpleDirectoryNode(qc, parentNo, pathSegments[0]!);
  }

  const pathAsJsonArray = JSON.stringify(pathSegments);

  const insertionPoint = await qc
    .withRecursive("path_segment", (ctec) =>
      ctec
        .selectFrom(sql`json_each(${pathAsJsonArray})`.as("pathp"))
        .select(sql<number>`row_number() over ()`.as("segment_no"))
        .select(sql<string>`"pathp"."value"`.as("segment")),
    )
    .withRecursive("ids", (ctec) =>
      ctec

        .selectFrom(["node"])
        .select("node.node_no")
        .select(sql<bigint>`-1`.as("node_no_parent"))
        // i is the _next_ segment number on an one-based sequence
        .select(sql<bigint>`1`.as("i"))
        .where("node.node_no", "=", parentNo)
        .unionAll((uac) =>
          uac
            .selectFrom("ids")
            .innerJoin("path_segment", "path_segment.segment_no", "ids.i")
            .innerJoin("node_member", (join) =>
              join
                .onRef("node_member.node_no_parent", "=", "ids.node_no")
                .onRef("node_member.name", "=", "path_segment.segment"),
            )
            .select(["node_member.node_no", "node_member.node_no_parent"])
            .select(sql<bigint>`"ids"."i" + 1`.as("i")),
        )
        .orderBy("i"),
    )
    .selectFrom("ids")
    .innerJoin("node", "node.node_no", "ids.node_no")
    .select(["ids.node_no", "ids.node_no_parent", "ids.i", "node.node_type"])
    .orderBy("i", "desc")
    .limit(1)
    .executeTakeFirst();

  if (insertionPoint == null) {
    throw makeDbfsError(DbfsErrorCode.ENOENT, "Parent node does not exist");
  }
  if (insertionPoint.node_no_parent === parentNo) {
    if (insertionPoint.node_type === NodeType.DIRECTORY) {
      return BigInt(insertionPoint.node_no);
    }
    throw makeDbfsError(DbfsErrorCode.EEXIST, "Node is not a directory");
  }
  if (insertionPoint.node_type !== NodeType.DIRECTORY) {
    throw makeDbfsError(DbfsErrorCode.EEXIST, "Parent node is not a directory");
  }
  if (!recursive && Number(insertionPoint.i) !== pathSegments.length) {
    throw makeDbfsError(DbfsErrorCode.ENOENT, "Parent node does not exist");
  }

  let currentParentNodeNo = insertionPoint.node_no;
  for (let i = Number(insertionPoint.i) - 1; i < pathSegments.length; ++i) {
    const currentSegment = pathSegments[i]!;
    const node_no = await dbfsMakeSimpleDirectoryNode(
      qc,
      currentParentNodeNo,
      currentSegment,
    );
    currentParentNodeNo = node_no;
  }

  return BigInt(currentParentNodeNo);
}

async function dbfsMakeSimpleDirectoryNode(
  qc: AppQueryCreator,
  node_no_parent: bigint,
  name: string,
): Promise<bigint> {
  const insertRx = await qc
    .insertInto("node")
    .values({ node_type: "D" })
    .executeTakeFirst();
  if (insertRx.insertId == null) {
    throw makeLogicError("Failed to create a directory node");
  }
  const node_no = BigInt(insertRx.insertId);
  await Promise.all([
    qc.insertInto("node_directory").values({ directory_no: node_no }).execute(),
    qc
      .insertInto("node_member")
      .values({ node_no, name, node_no_parent })
      .execute(),
  ]);
  return node_no;
}
