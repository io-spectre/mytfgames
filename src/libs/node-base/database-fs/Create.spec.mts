import { DatabaseSync } from "node:sqlite";

import { Kysely, Migrator } from "kysely";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import {
  AppDatabase,
  ViteMigrationProvider,
  WellKnownDirectory,
} from "$node-base/database";

import { NodeSqliteDialect } from "../database/kysely-support/NodeSqliteDialect.mjs";
import { dbfsMakeDirectoryNode } from "./Create.mjs";

describe("database-fs:Create", () => {
  let db: Kysely<AppDatabase>;

  beforeEach(async () => {
    db = new Kysely<AppDatabase>({
      dialect: new NodeSqliteDialect({
        database: new DatabaseSync(":memory:"),
      }),
      log: ["error"],
    });
    const migrator = new Migrator({
      db,
      provider: new ViteMigrationProvider(),
    });
    const migration = await migrator.migrateToLatest();
    if (migration.error != null) {
      throw new Error("Failed to migrate the database schema", {
        cause: migration.error,
      });
    }

    await db
      .insertInto("node")
      .values([
        { node_type: "D", node_no: 2001n },
        { node_type: "D", node_no: 2002n },
        { node_type: "D", node_no: 2003n },
        { node_type: "D", node_no: 2004n },
        { node_type: "D", node_no: 2005n },
        { node_type: "D", node_no: 2006n },
        { node_type: "D", node_no: 2007n },
      ])
      .execute();
    await db
      .insertInto("node_directory")
      .values([
        { directory_no: 2001n },
        { directory_no: 2002n },
        { directory_no: 2003n },
        { directory_no: 2004n },
        { directory_no: 2005n },
        { directory_no: 2006n },
        { directory_no: 2007n },
      ])
      .execute();
    await db
      .insertInto("node_member")
      .values([
        { node_no: 2001n, node_no_parent: WellKnownDirectory.TMP, name: "a" },
        { node_no: 2002n, node_no_parent: WellKnownDirectory.ROOT, name: "b" },
        { node_no: 2003n, node_no_parent: 2001n, name: "c" },
        { node_no: 2004n, node_no_parent: 2001n, name: "d" },
        { node_no: 2005n, node_no_parent: 2004n, name: "e" },
        { node_no: 2006n, node_no_parent: 2004n, name: "f" },
        { node_no: 2007n, node_no_parent: 2005n, name: "g" },
      ])
      .execute();
  });
  afterEach(() => {
    return db.destroy();
  });

  describe("dbfsMakeDirectoryNode", () => {
    it("creates a simple subdirectory node", async () => {
      const nodeNo = await dbfsMakeDirectoryNode(db, "test");
      expect(nodeNo).toBe(2008n);

      const node = await db
        .selectFrom("node")
        .selectAll()
        .where("node.node_no", "=", nodeNo)
        .execute();
      expect(node).toMatchObject([{ node_no: 2008, node_type: "D" }]);

      const member = await db
        .selectFrom("node_member")
        .selectAll()
        .where("node_member.node_no", "=", nodeNo)
        .execute();
      expect(member).toMatchObject([
        {
          node_no: 2008,
          node_no_parent: Number(WellKnownDirectory.ROOT),
          name: "test",
        },
      ]);
    });

    it.each([
      ["test", WellKnownDirectory.ROOT, { node_no: 2008, name: "test" }],
      ["test", WellKnownDirectory.TMP_IMPORT, { node_no: 2008, name: "test" }],
      [
        "tmp/t4",
        WellKnownDirectory.ROOT,
        {
          node_no: 2008,
          node_no_parent: Number(WellKnownDirectory.TMP),
          name: "t4",
        },
      ],
      [
        "import/t4",
        WellKnownDirectory.TMP,
        {
          node_no: 2008,
          node_no_parent: Number(WellKnownDirectory.TMP_IMPORT),
          name: "t4",
        },
      ],
    ])("creates a simple directory node", async (path, parent, expected) => {
      const nodeNo = await dbfsMakeDirectoryNode(db, parent, path);
      expect(nodeNo).toBe(BigInt(expected.node_no));

      const node = await db
        .selectFrom("node")
        .selectAll()
        .where("node.node_no", "=", nodeNo)
        .execute();
      expect(node).toMatchObject([
        { node_no: expected.node_no, node_type: "D" },
      ]);

      const member = await db
        .selectFrom("node_member")
        .selectAll()
        .where("node_member.node_no", "=", nodeNo)
        .execute();
      expect(member).toMatchObject([
        {
          node_no_parent: Number(parent),
          ...expected,
        },
      ]);
    });
    it.each([
      [
        "tx/t3",
        WellKnownDirectory.ROOT,
        { node_no: 2009, node_no_parent: 2008, name: "t3" },
      ],
      ["t2/t3", 2001n, { node_no: 2009, node_no_parent: 2008, name: "t3" }],
    ])("creates a deep directory structure", async (path, parent, expected) => {
      const nodeNo = await dbfsMakeDirectoryNode(db, parent, path, {
        recursive: true,
      });
      expect(nodeNo).toBe(BigInt(expected.node_no));

      const node = await db
        .selectFrom("node")
        .selectAll()
        .where("node.node_no", "=", nodeNo)
        .execute();
      expect(node).toMatchObject([
        { node_no: expected.node_no, node_type: "D" },
      ]);

      const member = await db
        .selectFrom("node_member")
        .selectAll()
        .where("node_member.node_no", "=", nodeNo)
        .execute();
      expect(member).toMatchObject([expected]);
    });
    it("throws an error if the parent node does not exist", async () => {
      await expect(
        dbfsMakeDirectoryNode(db, 9999n, "test"),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
