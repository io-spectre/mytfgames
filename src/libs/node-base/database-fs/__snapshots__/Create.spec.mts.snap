// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`database-fs:Create > dbfsMakeDirectoryNode > throws an error if the parent node does not exist 1`] = `[Error: Parent node does not exist]`;
